Added cached_write functionality

shaidar opened the pull request:
To use vault pki functionality and avoid generating certs when pillar is rendered, we need a way to cache certs and only generate new ones when current ones are about to expire. This is different than the other creds since this requires a `POST` request as opposed to a `GET` request to generate the cert and thus we decided to add a `cached_write` function to handle that request.

blarghmatey commented:
:+1:

blarghmatey commented:
For now we can leverage this via Jinja, but it may be useful to take a crack at updating the renderer parsing to allow for parameters to functions.

e.g. `__vault__:cache_write|subject=consul:operations-pki/creds/cert>data>key`

Also, with more thought on the matter, it might be useful to factor in the subject or other keywords as part of the cache path...

shaidar commented:
Well adding the `common_name` keyword to pass as a parameter would be useful

shaidar commented:
@blarghmatey This needs another review as I refactored code by moving similar functionality between `cached_read` and `cached_write` into its own function and also added a check to renew a secret if possible.

blarghmatey commented on _modules/vault.py:
```diff
@@ -177,16 +177,32 @@ def cached_read(path, cache_prefix='', **kwargs):
-            vault_client.delete(cache_path)
-
-    if not vault_data or not lease_valid:
+            if not vault_data['renewable']:
```
I would add an extra condition based on a keyword arg so we can override this behavior in the event that we don't want to renew leases for certain situations.

shaidar replied:
In looking at how you're handling `lease_valid` I'm wondering about the case when `if not vault_data or not lease_valid` [is met](https://github.com/mitodl/vault-formula/blob/master/_modules/vault.py#L182). Shouldn't that throw an undefined error for the `lease_valid` keyword since that gets defined only when `vault_data` is non-empty?

blarghmatey replied:
The `or` operator is a short-circuit logic path so it won't every try to evaluate `lease_valid` in the event that there is no vault data.

blarghmatey commented on _modules/vault.py:
```diff
@@ -177,16 +177,32 @@ def cached_read(path, cache_prefix='', **kwargs):
+
+
+def cached_read(path, cache_prefix='', **kwargs):
+    check_cached_lease(path, cache_prefix='', **kwargs)
```
This isn't assigning any of the return values to local variables, so `vault_data` in the next line will be undefined.

blarghmatey commented:
I think this looks good now :+1:
